File: amazon/5_longest_palindromic_substring.cpp
class Solution {
public:

    string transform(string s){                                                                                     
    string temp="#";                                                                                            
    for(char ch:s){                                                                                             
        temp+=ch;                                                                                               
        temp+="#";                                                                                              
    }                                                                                                           
    return temp;                                                                                                
}                                                                                                               
                                                                                                                
string longestpalindrome(string s){                                                                             
    string t= transform(s);                                                                                     
    int n=t.size();                                                                                             
    int l=0;   //to check the left boundary of the right most palindrome found so far                           
    int r=0;   //to check the right boundary of the right most palindrome found so far                          
    int center;  //to track the index of the longest palindrome                                                 
    int maxlen=0;  //to track the longest palindrome                                                            
    vector<int>p(n);                                                                                            
    p[0]=0;                                                                                                     
    for(int i=1;i<n;i++){                                                                                       
        int k;                                                                                                  
        //Expand around the ith index to compute p[i]                                                           
        if(i>r){                                                                                                
           k=0;                                                                                                 
        }                                                                                                       
        else{                                                                                                   
            int j=l+(r-i);                                                                                      
            if(j-p[j]>l){                                                                                       
                p[i]=p[j];                                                                                      
                continue;                                                                                       
            }                                                                                                   
            else{                                                                                               
                k=r-i;                                                                                          
            }                                                                                                   
        }                                                                                                       
        while (i-k>=0 && i+k<n && t[i-k]==t[i+k]){                                                              
            k++;                                                                                                
        }                                                                                                       
        k--;                                                                                                    
        p[i]=k;                                                                                                 
        if(p[i]>maxlen){                                                                                        
            maxlen=p[i];                                                                                        
            center=i/2;                                                                                         
        }                                                                                                       
        if(i+k>r){                                                                                              
            l=i-k;                                                                                              
            r=i+k;                                                                                              
        }                                                                                                       
    }                                                                                                           
    return s.substr(center-(maxlen/2),maxlen);                                                                  
}                                                                                                               
                                                                                                                

    string longestPalindrome(string s) {
        return longestpalindrome(s);
    }
};

File: amazon/5_longest_palindromic_substring.py
# Given a string s, return the longest palindromic substring in s.
# Example 1:
# Input: s = "babad"
# Output: "bab"
# Explanation: "aba" is also a valid answer.
# Example 2:
# Input: s = "cbbd"
# Output: "bb"
# Constraints:
# 1 <= s.length <= 1000
# s consist of only digits and English letters.
#expand around the center
    
class Solution:
    def longestPalindrome(self, s: str) -> str:
        if len(s) <=1:
            return s
        def expand_from_center(left, right):
            while left >= 0 and right < len(s) and s[left] == s[right]:
                left -= 1
                right += 1
            return s[left + 1:right]

        max_str = s[0]

        for i in range(len(s) -1):
            odd = expand_from_center(i, i)
            even = expand_from_center(i, i + 1)

            if len(odd) > len(max_str):
                max_str = odd 
            if len(even) > len(max_str):
                max_str = even

        return max_str

#memory friendly improvement
class Solution:
    def longestPalindrome(self, s: str)-> str:
        if s == s[::-1]:
            return s
        max_size = len(s)
        window_size = max_size -1
        while window_size > 1:
            for i in range(max_size - window_size +1):
                sub_str = s[i:window_size+i]
                if sub_str == sub_str[::-1]:
                    return sub_str
            window_size -= 1
    return s[0]


        
